lib/compiler/src/beam_doc.erl

```diff
@@ -0,0 +1,579 @@
+   Acc1 = extract_user_types(A, Acc),
+   extract_user_types(Args, Acc1);
+extract_user_types({type, _, bounded_fun, Args}, Acc) ->
+   extract_user_types(Args, Acc);
+extract_user_types({type, _, 'fun', Args}, Acc) ->
+   extract_user_types(Args, Acc);
+extract_user_types({type,_,product,Args}, Acc) ->
+   extract_user_types(Args, Acc);
+extract_user_types({type, _, map, Args}, Acc) ->
+   extract_user_types(Args, Acc);
+extract_user_types({type, _, _N, Args}, Acc) ->
+   extract_user_types(Args, Acc);
+extract_user_types({ann_type, _, [_Name, Type]}, Acc) ->
+   extract_user_types(Type, Acc);
+extract_user_types({type, _,record,[_Name | Args]}, Acc) ->
+   extract_user_types(Args, Acc);
+extract_user_types({type, _, tuple, Args}, Acc) ->
+   extract_user_types(Args, Acc);
+extract_user_types({type, _,union, Args}, Acc) ->
+   extract_user_types(Args, Acc);
+extract_user_types({type, _, map_field_assoc, Args}, Acc) ->
+   extract_user_types(Args, Acc);
+extract_user_types({type, _, map_field_exact, Args}, Acc) ->
+   extract_user_types(Args, Acc);
+extract_user_types({user_type, _, Name, Args}, Acc) ->
+   %% append user type and continue iterating through lists in case of other
+   %% user-defined types to be added
+   Acc1 = sets:add_element({Name, length(Args)}, Acc),
+   extract_user_types(Args, Acc1);
+extract_user_types(_Else, Acc) ->
+    Acc.
+
+
+
+%% Accumulate list of variables. if some type is not simply a var,
+%% i.e., distinct from foo(X, Y) where all arguments are variables,
+%% return false to signal that we cannot extract a slogan from them spec
+extract_slogan_from_spec({attribute, _Anno, Tag, Form}, State) when Tag =:= spec; Tag =:= callback ->
+   maybe
+      {Name, Arity, {type, _, 'fun', [{type, _, product, Args}, _Return]}} ?=
+         case Form of
+            {{Name0, Arity0}, Types} -> {Name0, Arity0, hd(Types)}; % hd(Types) to ignore the return type.
+            {{_Mod, Name0, Arity0}, Types} -> {Name0, Arity0, hd(Types)}
+         end,
+      Vars = foldl(fun (_, false) -> false;
+                             ({var, _, Var}, Vars) -> [Var | Vars];
+                             (_, _) -> false
+                         end, [], Args),
+      true ?= is_list(Vars),
+      Arity ?= length(Vars),
+      update_slogan(State, {Name, Vars, Arity})
+   else
+      _ ->
+         State
+   end.
+
+%% NOTE: Terminal elements for the documentation, such as `-type`, `-opaque`, `-callback`,
+%%       and functions always need to reset the state when they finish, so that new
+%%       new AST items start with a clean slate.
+extract_documentation_from_type([{attribute, Anno, TypeOrOpaque, {Type, _, TypeArgs}}=_AST | T],
+                      #docs{exported_types=ExpTypes, meta=Meta}=State)
+  when TypeOrOpaque =:= type; TypeOrOpaque =:= opaque ->
+    Args = fun_to_varargs(TypeArgs),
+    case sets:is_element({Type, length(Args)}, ExpTypes) of
+       true ->
+          State1 = State#docs{ meta = Meta#{exported := true}},
```

Comment:
Types can refer to private types that can refer to private types. So user types need to scan recursively through types. For example:

```erlang
-export([foo/1]).
-type one() :: 1.
-type two() :: one().
-type three() :: two().
-type four() :: three().

-spec foo() -> three().
foo() -> 1.
```

In the above, all the types except four/0 should be included in the docs.